packages/vat-data/src/vat-data-bindings.js

```diff
@@ -91,19 +93,115 @@ export const partialAssign = (target, source) => {
 // @ts-expect-error TODO statically recognize harden
 harden(partialAssign);
 
+/**
+ * @type {WeakMap<import('./types.js').Baggage, Map<string, (k?:string) => any>>}
+ */
+const collisions = new WeakMap();
+
+/**
+ * HAZARD: `assertUniqueProvide` uses top-level mutable state, but only for
+ * diagnostic purposes.
+ * Aside from being able to provoke it to throw an error, it is not
+ * otherwise usable as a top-level communications channel. The throwing-or-not
+ * of the error is the only way in which this top level state is observable.
+ *
+ * TODO since this still does violate our security assumptions, we should
+ * consider making it development-time only, and turn it off in production.
+ * Or better, adapt it to stop using top-level state.
+ *
+ * The condition it checks for is that, within one incarnation, the
+ * `provide` is never called twice with the same `mapStore`
+ * and the same `key`. For a given `mapStore` and `key`, there should be at
+ * most one `provide` call but any number of `get` calls.
+ *
+ * When the `mapStore` is baggage, for purposes of upgrade, then we
+ * expect `provide` to be called with the same `baggage` and `key`
+ * (and some `makeValue`) in each incarnation of that vat.
+ * In the incarnation that first registers this `baggage`,`key` pair, the
+ * `provide` will actually call the `makeValue`. The one `provide` call
+ * for this pair in each succeeding incarnation will merely retrieve the value it
+ * inherited from the previous incarnation. But even for this case, it should
+ * only call `provide` for that pair once during that incarnation.
+ *
+ * @param {import('./types.js').Baggage} baggage
+ * @param {string} key
+ * @param {(key?: string) => unknown} makeValue
+ */
+const assertUniqueProvide = (baggage, key, makeValue) => {
+  if (collisions.has(baggage)) {
+    const submap = /** @type {Map<string, (k?:string) => any>} */ (
+      collisions.get(baggage)
+    );
+    if (submap.has(key)) {
+      if (submap.get(key) === makeValue) {
+        // The error message also tells us if the colliding calls happened
+        // to use the same `makeValue` only as a clue about whether
+        // these are two separate evaluations at of the same call site or
+        // two distinct call sites. However, this clue is unreliable in
+        // both directions.
+        assert.fail(X`provide collision on ${q(key)} with same makeValue`);
```

Comment:
We can help the developer solve this problem:
```suggestion
        assert.fail(X`provide collision on ${q(key)} with same makeValue. Consider provideLazy()`);
```

Reply:
Done